File: android-native/app/src/main/java/com/example/kpkn/screens/wikilab/WikiLabScreen.kt
package com.example.kpkn.screens.wikilab

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.LazyRow
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.lazy.rememberLazyListState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.example.kpkn.data.exercises.exerciseCatalogSnapshot
import com.example.kpkn.data.models.ExerciseMuscleInfo
import com.example.kpkn.data.models.MuscleRole
import com.example.kpkn.data.repository.CustomExerciseRepository
import com.example.kpkn.domain.exercises.calculateSearchScore

// ═══════════════════════════════════════════════════════════════════════
// CATEGORY FILTER
// ═══════════════════════════════════════════════════════════════════════

private data class MuscleCategory(val label: String, val keywords: List<String>)

private val CATEGORIES = listOf(
    MuscleCategory("Todos", emptyList()),
    MuscleCategory("Pecho", listOf("Pectorales")),
    MuscleCategory("Espalda", listOf("Dorsales", "Trapecio", "Erectores Espinales")),
    MuscleCategory("Hombros", listOf("Deltoides")),
    MuscleCategory("Piernas", listOf("Cuádriceps", "Isquiosurales", "Glúteos", "Pantorrillas", "Aductores")),
    MuscleCategory("Brazos", listOf("Bíceps", "Tríceps", "Antebrazo")),
    MuscleCategory("Core", listOf("Core", "Abdomen")),
)

// ═══════════════════════════════════════════════════════════════════════
// MAIN COMPOSABLE
// ═══════════════════════════════════════════════════════════════════════

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun WikiLabScreen(
    modifier: Modifier = Modifier,
    onOpenExercise: (String) -> Unit,
    onBack: () -> Unit = {},
) {
    var query by remember { mutableStateOf("") }
    var selectedCategory by remember { mutableIntStateOf(0) }

    val customExercises by CustomExerciseRepository.customExercises.collectAsState()
    val exerciseCatalog = remember(customExercises) {
        (exerciseCatalogSnapshot() + customExercises)
            .associateBy { it.id.lowercase() }
            .values
            .toList()
    }

    val filtered = remember(query, selectedCategory, exerciseCatalog) {
        val cat = CATEGORIES[selectedCategory]
        val matches = exerciseCatalog.filter { ex ->
            val canonicalInvolved = collapseInvolvedMusclesToCanonical(ex.involvedMuscles)
            val catMatch = cat.keywords.isEmpty() || ex.involvedMuscles.any { m ->
                cat.keywords.any { kw -> canonicalMuscleDisplayName(m.muscle, m.emphasis).contains(kw, ignoreCase = true) }
            }
            val textMatch = if (query.isBlank()) {
                true
            } else {
                calculateSearchScore(ex, query) > 0 ||
                    canonicalInvolved.any { m -> m.muscle.contains(query, ignoreCase = true) }
            }
            catMatch && textMatch
        }
        val sorted = if (query.isBlank()) {
            matches.sortedBy { it.name }
        } else {
            matches.sortedWith(
                compareByDescending<ExerciseMuscleInfo> { calculateSearchScore(it, query) }
                    .thenBy { kotlin.math.abs(it.name.length - query.trim().length) }
                    .thenBy { it.name }
            )
        }
        // Aprende is an index of the approved definitions. Do not apply the
        // session-editor's visual name deduplication here: it can collapse
        // distinct catalog definitions that happen to share aliases. The
        // canonical id remains the sole identity for this list.
        sorted.distinctBy { it.id.lowercase() }
    }
    val listState = rememberLazyListState()
    LaunchedEffect(query, selectedCategory) {
        listState.scrollToItem(0)
    }

    Scaffold(
        containerColor = APRENDE_BACKGROUND,
        topBar = {}, // Cabecera fija eliminada para máximo espacio de scroll
    ) { paddingValues ->
        Box(
            modifier = modifier
                .fillMaxSize()
                .background(APRENDE_BACKGROUND)
                .padding(paddingValues)
        ) {
            LazyColumn(
                state = listState,
                modifier = Modifier.fillMaxSize(),
                contentPadding = PaddingValues(horizontal = 20.dp, vertical = 12.dp),
            ) {
                // ─── Scrollable Header (Retroceso + Título) ───────────────
                item {
                    Row(
                        modifier = Modifier
                            .fillMaxWidth()
                            .statusBarsPadding()
                            .padding(vertical = 8.dp),
                        verticalAlignment = Alignment.CenterVertically
                    ) {
                        IconButton(onClick = onBack) {
                            Icon(Icons.Default.ArrowBack, contentDescription = "Volver", tint = Color.White)
                        }
                        Spacer(Modifier.width(8.dp))
                        Text(
                            text = "Aprende · ejercicios",
                            style = MaterialTheme.typography.headlineLarge.copy(fontFamily = FontFamily.Serif),
                            fontWeight = FontWeight.Black,
                            color = Color.White,
                        )
                    }
                }

                // ─── Subtitle & Description ───────────────────────────────
                item {
                    Column(verticalArrangement = Arrangement.spacedBy(6.dp)) {
                        Text(
                            text = "Índice editorial basado en el catálogo aprobado. Abre un ejercicio para consultar su variante exacta, anatomía y técnica.",
                            style = MaterialTheme.typography.bodyMedium.copy(fontFamily = FontFamily.Serif),
                            color = Color.White.copy(alpha = 0.7f),
                        )
                        Spacer(Modifier.height(4.dp))
                    }
                }

                // ─── Search Bar ───────────────────────────────────────────
                item {
                    OutlinedTextField(
                        value = query,
                        onValueChange = { query = it },
                        modifier = Modifier.fillMaxWidth(),
                        placeholder = { Text("Buscar ejercicio...", style = MaterialTheme.typography.bodySmall.copy(fontFamily = FontFamily.Serif)) },
                        leadingIcon = { Icon(Icons.Default.Search, null) },
                        trailingIcon = {
                            if (query.isNotEmpty()) {
                                IconButton(onClick = { query = "" }) {
                                    Icon(Icons.Default.Close, "Limpiar")
                                }
                            }
                        },
                        singleLine = true,
                        shape = RoundedCornerShape(4.dp),
                        colors = OutlinedTextFieldDefaults.colors(
                            focusedTextColor = Color.White,
                            unfocusedTextColor = Color.White,
                            focusedBorderColor = Color.White.copy(alpha = 0.4f),
                            unfocusedBorderColor = Color.White.copy(alpha = 0.2f),
                            cursorColor = Color.White,
                        ),
                    )
                    Spacer(Modifier.height(8.dp))
                }

                // ─── Category Filter Chips ────────────────────────────────
                item {
                    LazyRow(
                        horizontalArrangement = Arrangement.spacedBy(8.dp),
                        contentPadding = PaddingValues(vertical = 4.dp),
                    ) {
                        items(CATEGORIES.indices.toList()) { idx ->
                            val cat = CATEGORIES[idx]
                            val selected = selectedCategory == idx
                            FilterChip(
                                selected = selected,
                                onClick = { selectedCategory = idx },
                                label = {
                                    Text(
                                        cat.label,
                                        style = MaterialTheme.typography.labelMedium,
                                        fontWeight = if (selected) FontWeight.Black else FontWeight.SemiBold,
                                    )
                                },
                                colors = FilterChipDefaults.filterChipColors(
                                    containerColor = Color.Transparent,
                                    selectedContainerColor = APRENDE_MUTED_FILL,
                                    labelColor = Color.White.copy(alpha = 0.72f),
                                    selectedLabelColor = Color.White,
                                    iconColor = APRENDE_LINK_COLOR,
                                    selectedLeadingIconColor = APRENDE_LINK_COLOR,
                                ),
                                leadingIcon = if (idx > 0) {
                                    {
                                    Surface(Modifier.size(6.dp), CircleShape, Color(0xFF7F8D96)) {}
                                    }
                                } else null,
                                shape = RoundedCornerShape(4.dp),
                            )
                        }
                    }
                    Spacer(Modifier.height(10.dp))
                }

                // ─── Results Counter ──────────────────────────────────────
                item {
                    Text(
                        text = "${filtered.size} artículos",
                        style = MaterialTheme.typography.labelSmall.copy(fontFamily = FontFamily.Serif),
                        color = Color.White.copy(alpha = 0.5f),
                        modifier = Modifier.padding(vertical = 4.dp),
                    )
                    Spacer(Modifier.height(6.dp))
                }

                // ─── List of Exercise Rows ────────────────────────────────
                items(filtered, key = { it.id }) { exercise ->
                    ExerciseCard(exercise = exercise, onClick = { onOpenExercise(exercise.id) })
                }

                item { Spacer(Modifier.height(80.dp)) }
            }
        }
    }
}// ═══════════════════════════════════════════════════════════════════════
// EXERCISE ROW (NO CARDS)
// ═══════════════════════════════════════════════════════════════════════

@Composable
private fun ExerciseCard(exercise: ExerciseMuscleInfo, onClick: () -> Unit) {
    val canonicalInvolved = remember(exercise.id, exercise.involvedMuscles) {
        collapseInvolvedMusclesToCanonical(exercise.involvedMuscles)
    }
    val primaryMuscles = canonicalInvolved.filter { it.role == MuscleRole.PRIMARY }
    val secondaryCount = canonicalInvolved.count { it.role != MuscleRole.PRIMARY }
    Column(
        modifier = Modifier
            .fillMaxWidth()
            .clickable(onClick = onClick)
            .padding(vertical = 12.dp)
    ) {
        Text(
            text = exercise.name,
            style = MaterialTheme.typography.bodyLarge.copy(
                fontFamily = FontFamily.Serif,
                color = APRENDE_LINK_COLOR
            ),
            fontWeight = FontWeight.Bold,
        )
        
        Spacer(Modifier.height(4.dp))
        
        if (primaryMuscles.isNotEmpty()) {
            Text(
                text = "Músculos principales: " + primaryMuscles.joinToString(", ") { it.muscle } +
                    (if (secondaryCount > 0) " (+$secondaryCount secundarios)" else ""),
                style = MaterialTheme.typography.bodyMedium.copy(fontFamily = FontFamily.Serif),
                color = Color.White.copy(alpha = 0.8f),
            )
        } else if (secondaryCount > 0) {
            Text(
                text = "$secondaryCount músculos secundarios",
                style = MaterialTheme.typography.bodyMedium.copy(fontFamily = FontFamily.Serif),
                color = Color.White.copy(alpha = 0.8f),
            )
        }

        exercise.equipment?.takeIf { it.isNotBlank() }?.let { equipment ->
            Text(
                text = "Equipamiento: $equipment",
                style = MaterialTheme.typography.labelSmall.copy(fontFamily = FontFamily.Serif),
                color = Color.White.copy(alpha = 0.5f),
            )
        }
        
        Spacer(Modifier.height(10.dp))
        HorizontalDivider(color = APRENDE_DIVIDER, thickness = 1.dp)
    }
}

File: ios-native/KPKNFit/KPKNFit/Presentation/Screens/Wikilab/WikiLabScreen.swift
import SwiftUI

private struct MuscleCategory {
    let label: String
    let keywords: [String]
    let color: Color
}

private let CATEGORIES: [MuscleCategory] = [
    MuscleCategory(label: "Todos", keywords: [], color: Color(hex: 0x8A9099)),
    MuscleCategory(label: "Pecho", keywords: ["Pectorales"], color: Color(hex: 0x1E88E5)),
    MuscleCategory(label: "Espalda", keywords: ["Dorsales", "Trapecio", "Erectores Espinales"], color: Color(hex: 0x43A047)),
    MuscleCategory(label: "Hombros", keywords: ["Deltoides"], color: Color(hex: 0xFFFF8F00)),
    MuscleCategory(label: "Piernas", keywords: ["Cuádriceps", "Isquiosurales", "Glúteos", "Pantorrillas", "Aductores"], color: Color(hex: 0x9C27B0)),
    MuscleCategory(label: "Brazos", keywords: ["Bíceps", "Tríceps", "Antebrazo"], color: Color(hex: 0xE53935)),
    MuscleCategory(label: "Core", keywords: ["Core", "Abdomen"], color: Color(hex: 0x00ACC1)),
]

struct WikiLabScreen: View {
    let onCreateExercise: () -> Void
    let onOpenExercise: (String) -> Void
    let onBack: () -> Void
    
    @State private var query: String = ""
    @State private var selectedCategory: Int = 0
    
    private var customExercises: [ExerciseMuscleInfo] { CustomExerciseRepository.shared.customExercises }
    
    private var exerciseCatalog: [ExerciseMuscleInfo] {
        let all = catalogExerciseList + customExercises
        var dict = [String: ExerciseMuscleInfo]()
        for ex in all { dict[ex.id.lowercased()] = ex }
        return Array(dict.values)
    }
    
    private var filtered: [ExerciseMuscleInfo] {
        let cat = CATEGORIES[selectedCategory]
        let matches = exerciseCatalog.filter { ex in
            let canonicalInvolved = collapseInvolvedMusclesToCanonical(ex.involvedMuscles)
            let catMatch = cat.keywords.isEmpty || ex.involvedMuscles.contains { m in
                cat.keywords.contains { kw in
                    canonicalMuscleDisplayName(m.muscle, emphasis: m.emphasis).localizedCaseInsensitiveContains(kw)
                }
            }
            let textMatch: Bool
            if query.trimmingCharacters(in: .whitespaces).isEmpty {
                textMatch = true
            } else {
                textMatch = calculateSearchScore(ex, query) > 0 ||
                    canonicalInvolved.contains { $0.muscle.localizedCaseInsensitiveContains(query) }
            }
            return catMatch && textMatch
        }
        let sorted: [ExerciseMuscleInfo]
        if query.trimmingCharacters(in: .whitespaces).isEmpty {
            sorted = matches.sorted { $0.name < $1.name }
        } else {
            sorted = matches.sorted { a, b in
                let sa = calculateSearchScore(a, query)
                let sb = calculateSearchScore(b, query)
                if sa != sb { return sa > sb }
                let da = abs(a.name.count - query.trimmingCharacters(in: .whitespaces).count)
                let db = abs(b.name.count - query.trimmingCharacters(in: .whitespaces).count)
                if da != db { return da < db }
                return a.name < b.name
            }
        }
        return deduplicateCatalogVisualResults(sorted)
    }
    
    var body: some View {
        ZStack {
            Color.black.ignoresSafeArea()
            
            ScrollViewReader { proxy in
                List {
                    headerSection
                    subtitleSection
                    searchBarSection
                    categoryFilterSection
                    resultsCountSection
                    exerciseListSection
                    Color.clear.frame(height: 80).id("bottom-spacer")
                }
                .listStyle(.plain)
                .background(Color.black)
                .scrollContentBackground(.hidden)
                .onChange(of: query) { _ in
                    withAnimation { proxy.scrollTo("top", anchor: .top) }
                }
                .onChange(of: selectedCategory) { _ in
                    withAnimation { proxy.scrollTo("top", anchor: .top) }
                }
            }
            
            VStack {
                Spacer()
                HStack {
                    Spacer()
                    Button(action: onCreateExercise) {
                        Image(systemName: "plus")
                            .font(.title2.weight(.semibold))
                            .foregroundColor(.black)
                            .frame(width: 56, height: 56)
                            .background(AppColors.neonCyan)
                            .clipShape(RoundedRectangle(cornerRadius: 8))
                    }
                    .padding(16)
                }
            }
        }
        .navigationBarHidden(true)
    }
    
    private var headerSection: some View {
        HStack(spacing: 8) {
            Button(action: onBack) {
                Image(systemName: "chevron.left")
                    .font(.title3)
                    .foregroundColor(.white)
            }
            Text("Ejercicios")
                .font(.system(size: 32, weight: .black, design: .serif))
                .foregroundColor(.white)
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .padding(.vertical, 8)
        .listRowInsets(EdgeInsets())
        .listRowBackground(Color.black)
        .id("top")
    }
    
    private var subtitleSection: some View {
        Text("Índice de ejercicios y guía anatómica. Selecciona un elemento para ver su nivel, músculos implicados, equipamiento y fatiga.")
            .font(.system(size: 14, design: .serif))
            .foregroundColor(.white.opacity(0.7))
            .listRowInsets(EdgeInsets())
            .listRowBackground(Color.black)
    }
    
    private var searchBarSection: some View {
        HStack(spacing: 8) {
            Image(systemName: "magnifyingglass")
                .foregroundColor(.white.opacity(0.5))
            if query.isEmpty {
                Text("Buscar artículo...")
                    .font(.system(size: 13, design: .serif))
                    .foregroundColor(.white.opacity(0.3))
                Spacer()
            } else {
                TextField("", text: $query)
                    .font(.system(size: 13, design: .serif))
                    .foregroundColor(.white)
                    .autocorrectionDisabled()
                Button(action: { query = "" }) {
                    Image(systemName: "xmark")
                        .foregroundColor(.white.opacity(0.5))
                }
            }
        }
        .padding(12)
        .overlay(RoundedRectangle(cornerRadius: 4).stroke(Color.white.opacity(query.isEmpty ? 0.2 : 0.4), lineWidth: 1))
        .listRowInsets(EdgeInsets())
        .listRowBackground(Color.black)
        .padding(.vertical, 4)
    }
    
    private var categoryFilterSection: some View {
        ScrollView(.horizontal, showsIndicators: false) {
            HStack(spacing: 8) {
                ForEach(CATEGORIES.indices, id: \.self) { idx in
                    let cat = CATEGORIES[idx]
                    let isSelected = selectedCategory == idx
                    Button(action: { selectedCategory = idx }) {
                        HStack(spacing: 4) {
                            if idx > 0 {
                                Circle()
                                    .fill(cat.color)
                                    .frame(width: 6, height: 6)
                            }
                            Text(cat.label)
                                .font(.system(size: 12, weight: isSelected ? .black : .semibold))
                        }
                        .padding(.horizontal, 12)
                        .padding(.vertical, 6)
                        .background(isSelected ? Color.white.opacity(0.1) : Color.clear)
                        .overlay(RoundedRectangle(cornerRadius: 4).stroke(Color.white.opacity(0.2), lineWidth: 1))
                        .clipShape(RoundedRectangle(cornerRadius: 4))
                    }
                    .buttonStyle(.plain)
                    .foregroundColor(isSelected ? .white : .white.opacity(0.7))
                }
            }
            .padding(.vertical, 4)
        }
        .listRowInsets(EdgeInsets())
        .listRowBackground(Color.black)
    }
    
    private var resultsCountSection: some View {
        Text("\(filtered.count) artículos")
            .font(.system(size: 11, design: .serif))
            .foregroundColor(.white.opacity(0.5))
            .listRowInsets(EdgeInsets())
            .listRowBackground(Color.black)
            .padding(.vertical, 4)
    }
    
    private var exerciseListSection: some View {
        ForEach(filtered, id: \.id) { exercise in
            ExerciseCardView(exercise: exercise)
                .contentShape(Rectangle())
                .onTapGesture { onOpenExercise(exercise.id) }
                .listRowInsets(EdgeInsets())
                .listRowBackground(Color.black)
                .listRowSeparator(.hidden)
        }
    }
}

private struct ExerciseCardView: View {
    let exercise: ExerciseMuscleInfo
    
    private var canonicalInvolved: [InvolvedMuscle] {
        collapseInvolvedMusclesToCanonical(exercise.involvedMuscles)
    }
    
    private var primaryMuscles: [InvolvedMuscle] {
        canonicalInvolved.filter { $0.role == .PRIMARY }
    }
    
    private var secondaryCount: Int {
        canonicalInvolved.filter { $0.role != .PRIMARY }.count
    }
    
    private var equipment: String {
        exercise.equipment ?? "Peso Corporal"
    }
    
    var body: some View {
        VStack(alignment: .leading, spacing: 4) {
            Text(exercise.name)
                .font(.system(size: 16, weight: .bold, design: .serif))
                .foregroundColor(Color(hex: 0x29B6F6))
            
            Text("Músculos principales: " + primaryMuscles.map(\.muscle).joined(separator: ", ") + (secondaryCount > 0 ? " (+\(secondaryCount) secundarios)" : ""))
                .font(.system(size: 14, design: .serif))
                .foregroundColor(.white.opacity(0.8))
            
            Text("Equipamiento: \(equipment)")
                .font(.system(size: 11, design: .serif))
                .foregroundColor(.white.opacity(0.5))
            
            Divider()
                .background(Color(hex: 0x1A1A1A))
                .padding(.top, 6)
        }
        .padding(.vertical, 12)
    }
}
